controlpanel/frontend/consumers.py

```diff
@@ -1,59 +1,156 @@
+
+        # schedule a coroutine to send keepalive updates
+        asyncio.get_event_loop().create_task(self.stream())
+
+        # listen for messages for the current request user only
+        group = sanitize_dns_label(self.scope.get("user").auth0_id)
+        await self.channel_layer.group_add(group, self.channel_name)
+
+    async def stream(self):
+        """
+        Send a keepalive message every minute to prevent timeouts
+        """
+        while self.streaming:
+            await self.sse_event({
+                "event": "keepalive",
+                "data": datetime.now().isoformat(),
+            })
+            await asyncio.sleep(60)
+
+    async def sse_event(self, event):
+        """
+        Receive messages with {'type': 'sse.event'} and send as an SSE to the
+        client
+        """
+        payload = "\n".join(f"{key}: {value}" for key, value in event.items())
+
+        await self.send_body(
+            f"{payload}\n\n".encode("utf-8"),
+            more_body=self.streaming,
+        )
+
+    async def disconnect(self):
+        """
+        Stop the coroutine running on client disconnect
+        """
+        self.streaming = False
+
+        # leave the group
+        group = sanitize_dns_label(self.scope.get("user").auth0_id)
+        await self.channel_layer.group_discard(group, self.channel_name)
+
+
+class BackgroundTaskConsumer(SyncConsumer):
+
+    def tool_deploy(self, message):
+        """
+        Deploy the named tool for the specified user
+        Expects a message with `tool_name` and `user_id` values, eg:
+        message = {'tool_name': 'rstudio', 'user_id': 'github|1234'}
+        """
         tool = Tool.create(message['tool_name'])
         user = User.objects.get(auth0_id=message['user_id'])
 
-        log.debug(f'deploy_tool({tool.name}, {user.id})')
+        send_tool_status_update(user.auth0_id, tool.name, 'Deploying...')
 
         try:
             deploy = tool.deploy_for(user)
 
         except HelmError as err:
-            send_tool_status_change(tool.name, str(err))
-            log.error(err)
+            send_tool_status_update(user.auth0_id, tool.name, f"Error: {str(err)}")
+            raise err
 
         while deploy.poll() is None:
+            send_tool_status_update(user.auth0_id, tool.name, "Deploying...")
             sleep(1)
 
         if deploy.returncode:
             err_msg = deploy.stderr.read()
-            send_tool_status_change(tool.name, err_msg)
+            send_tool_status_update(user.auth0_id, tool.name, err_msg)
             log.error(err_msg)
             return
 
-        out = deploy.stdout.read()
-        send_tool_status_change(tool.name, out)
-        log.debug(out)
+        send_tool_status_update(user.auth0_id, tool.name, "Ready")
 
-    def restart_tool(self, message):
+    def tool_restart(self, message):
         tool = Tool.create(message['tool_name'])
         user = User.objects.get(auth0_id=message['user_id'])
```

Comment:
Apologies if I missed it, but how does authorisation works? We trust the incoming message and take the user ID from the message. I'm not familiar with this Python Channel thing so possibly I'm missing something.

Reply:
I suppose to send a message to the channel you need to talk with Redis.

Reply:
The channels router has authentication middleware wrapping the whole thing here:
https://github.com/ministryofjustice/analytics-platform-control-panel/pull/724/files#diff-310d0aa3904afc2adf8ab19949784881R14

Reply:
and yes, you can't talk directly to Redis from outside the cluster

Reply:
Isn't the authentication middleware only running for HTTP requests? I'm not familiar with Python Channels stuff (I'm having a read at their docs), maybe you don't need authorisation? 🤷‍♂

Reply:
I suppose the worst case scenario is an unauthorised user restart or deploy someone's tool, not the end of the world.

Reply:
good point, the authentication middleware is only for http, but there is no exposed endpoint/port for the background-tasks channel - as i understand it, you have to write to Redis to send a message on the channel, and Redis is not exposed outside the cluster

Reply:
Yes that's the conclusion I got to as well by reading this and some of the Django Channels documentation. I guess you're locking down Redis and that should be enough 🤷‍♂